### DISCUSSION TITLE
Combination between `insert_conflict` and `validate`

### DISCUSSION THREAD

**Discussion by jvanderen1** [jeremyevans/sequel#2177] [General]
## ⚠️ Issue

I am using the `:insert_conflict` Sequel model plugin and I believe there may be some difficulty using `insert_conflict` with `validate`.

For example, says  I have the following example model:

```ruby
# frozen_string_literal: true

# Table: users
# ---------------------------------------------------------------------------------------------------------
# Columns:
#  id                         | integer                     | PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY
#  email                      | text                        |
#  phone_number               | text                        |
# ---------------------------------------------------------------------------------------------------------

class User < Sequel.Model($db[:users])
  def validate
    super
    validates_unique :email
  end
end
```

Say the following `User` record exists:

```ruby
User.first.to_hash
=> {:id=>1,
 :email=>"foo@gmail.com",
 :phone_number=>"1234567890"}
```

I would like to update a `User` record if there is a conflicting value for `email` in my case. For example:

```ruby
target_hash = {
  email: 'foo@gmail.com'
}

update_hash = {
  phone_number: "0987654321"
}

User.new(**target_hash, **update_hash).
  insert_conflict(target: target_hash.keys, update: update_hash).
  save_changes
```

This unfortunately throws a `Sequel::ValidationFailed` error:

```sh
Sequel::ValidationFailed: email is already taken
```

Is there any good way to avoid this knowing that we are ignoring the uniqueness of `email` in this case via `insert_conflict`?

## Workaround 1: Skip Validate In Specific Cases (like `new?`)

A workaround I came up with is to define the `User` model `validate` method as the following instead:

```ruby
class User < Sequel.Model($db[:users])
  def validate
    super
    validates_unique :email unless new?
  end
end
``` 

However, this is not ideal as this workaround skips validation regardless of `insert_conflict` being present or not 😦 .

## Workaround 2: Disable Validation in `save_changes`

We could also do this:

```ruby
User.new(**target_hash, **update_hash).
  insert_conflict(target: target_hash.keys, update: update_hash).
  save_changes(validate: false)
```

However, we would then miss out on other forms of validation unfortunately.

**Comment by janko** [comment]
Seems like it would be nice if the `insert_conflict` plugin could let you know when it has been activated. Currently, you should be able to do:

```rb
validates_unique :email unless @insert_conflict_opts
```

But that seems like accessing private API of the plugin. Maybe if it provided a method like:

```rb
def insert_conflict?
  !!@insert_conflict_opts
end
```

**Reply by jvanderen1** [reply]
That would be great. I would love for it to be smart enough to automatically detect unique containt values with insert_conflict, but that may be difficult.
